kernel/task/src/lib.rs

```diff
@@ -409,14 +411,31 @@ impl JoinableTaskRef {
     ///   * This does *not* include cases where this `Task` failed or was killed,
     ///     rather only cases where the `join` operation itself failed.
     #[doc(alias("reap", "exit"))]
-    pub fn join(&self) -> Result<ExitValue, &'static str> {
-        // First, wait for this Task to be marked as Exited (no longer runnable).
-        while !self.has_exited() { }
-
-        // Then, wait for it to actually stop running on any CPU core.
-        while self.is_running() { }
-
-        // This synchronizes with the release fence from when this task first ran
+    pub fn join(&self) -> Result<ExitValue, &'static str> {        
+        if !self.has_exited() {
+            // Create a waker+blocker pair that will block the current task
+            // and then wake it once this task (`self`) exits.
+            let curr_task = get_my_current_task().ok_or("join(): couldn't get current task")?;
+            let task_to_block = curr_task.clone();
+            let wake_action = move || {
+                let _ = curr_task.unblock();
+            };
+            let (waker, blocker) = waker_generic::new_waker(wake_action);
+            self.set_waker(waker);
+            let block_action = || {
+                let _ = task_to_block.block();
+                ScheduleOnDrop { }
+            };
+            blocker.block(block_action);
+        }
+        
+        // @tsoutsman: double-check my reasoning in the comments below.
+        //
+        // Note: previously, we waited for this task to actually stop running;
+        //       however, I don't think there's any need to actually do that.
+        // while self.is_running() { }
```

Comment:
```rust
        // Note: previously, we waited for this task to actually stop running;
        //       however, I don't think there's any need to actually do that.
```
Is this the reasoning?

I don't think the fence will synchronise temporally with the fence in `task_wrapper`. We will be woken before `task_cleanup_final` has run but the exit value has already been stored so IDK if we really care.

Reply:
Yeah, basically that's the logic. Since we're only waiting for it to exit, we actually only care about the user-provided entry function being complete, _not_ our `task_wrapper` & friends being complete.

So yea, i think this is okay -- the various `task_cleanup*` functions can continue executing even after that task has been `join`ed by another task.

Reply:
> I don't think the fence will synchronise temporally with the fence in `task_wrapper`. We will be woken before `task_cleanup_final` has run but the exit value has already been stored so IDK if we really care.

Good point. We can leave the fence in `task_cleanup_final()` there as a backup in case wakers aren't used or the task isn't joined. But I will also put one in `TaskRef::internal_exit()` right before the waker is woken up.

Side note, i think we have too many useless fences, lol. But I'm not gonna worry about that.

Reply:
See new review comment.